### PR TITLE
gh-84436 Add integration C API tests for immortal objects

### PR DESCRIPTION
**Pull request by corona10**
Got motivation from https://github.com/python/cpython/pull/103922#issuecomment-1525895247
This PR is not the answer to the need for emulation mode but for minimum verification.

* Issue: gh-84436

### CODE REVIEW

File: PCbuild/_testcapi.vcxproj [hunk 1]
```diff
@@ -127,4 +128,4 @@
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
   <ImportGroup Label="ExtensionTargets">
   </ImportGroup>
-</Project>
\ No newline at end of file
+</Project>
```
Comment (corona10) [inline PCbuild/_testcapi.vcxproj]:
```suggestion
</Project>

```

File: Modules/_testcapi/immortal.c [hunk 2]
```diff
@@ -0,0 +1,105 @@
+#include "parts.h"
+
+static PyObject*
+test_immortal_bool(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    PyObject* objects[] = {Py_True, Py_False};
+    Py_ssize_t n = sizeof(objects) / sizeof(objects[0]);
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]: We actually have a macro for this in the C API ;) IIRC, it is `Py_ARRAY_SIZE`

File: Modules/_testcapi/immortal.c [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    for(Py_ssize_t i = 0; i < n; i++) {
+        PyObject* obj = objects[i];
+        if (!_Py_IsImmortal(obj)) {
+            PyErr_Format(PyExc_RuntimeError, "%R should be the immportal object.", obj);
+            return NULL;
+        }
+        Py_ssize_t old_count = Py_REFCNT(obj);
+        for (int i = 0; i < 10000; i++) {
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]: Use `j` iso. `i` for the loop counter

File: Modules/_testcapi/immortal.c [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            return NULL;
+        }
+    }
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+test_immortal_none(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    PyObject* none = Py_None;
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]: No need for a local variable. Just use `Py_None`

File: Modules/_testcapi/immortal.c [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+test_immortal_small_ints(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    for(int i = -5; i < 255; i++) {
+        PyObject *small_int = PyLong_FromLong(i);
+        if (!_Py_IsImmortal(small_int)) {
+            PyErr_Format(PyExc_RuntimeError, "Small int(%d) object should be the immportal object.", i);
+            return NULL;
+        }
+        Py_ssize_t old_count = Py_REFCNT(small_int);
+        for (int i = 0; i < 10000; i++) {
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]: Use different loop counter here also

File: Modules/_testcapi/immortal.c [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            return NULL;
+        }
+    }
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+test_immortal_ellipsis(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    PyObject* ellipsis = Py_Ellipsis;
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]: Ditto: no need for a local variable; just use `Py_Ellipsis`

**Review by erlend-aasland — COMMENTED** [review]
I took a quick look on my phone. I'll take a better look later today.

File: Modules/_testcapi/immortal.c (hunk 3 above)
Comment (corona10) [reply Modules/_testcapi/immortal.c]: Oops.

File: Modules/_testcapi/immortal.c [hunk 7]
```diff
@@ -0,0 +1,103 @@
+#include "parts.h"
+
+static PyObject*
+test_immortal_bool(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    PyObject* objects[] = {Py_True, Py_False};
+    Py_ssize_t n = Py_ARRAY_LENGTH(objects);
+    for(Py_ssize_t i = 0; i < n; i++) {
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]:
```suggestion
    for (Py_ssize_t i = 0; i < n; i++) {
```

File: Modules/_testcapi/immortal.c [hunk 8]
```diff
@@ -0,0 +1,103 @@
+#include "parts.h"
+
+static PyObject*
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]:
```suggestion
static PyObject *
```

File: Modules/_testcapi/immortal.c [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+test_immortal_bool(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    PyObject* objects[] = {Py_True, Py_False};
+    Py_ssize_t n = Py_ARRAY_LENGTH(objects);
+    for(Py_ssize_t i = 0; i < n; i++) {
+        PyObject* obj = objects[i];
+        if (!_Py_IsImmortal(obj)) {
+            PyErr_Format(PyExc_RuntimeError, "%R should be the immportal object.", obj);
+            return NULL;
+        }
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]: It's ok to use asserts in the test CAPI C tests.

```suggestion
        assert(_Py_IsImmortal(obj));
```

File: Modules/_testcapi/immortal.c [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        Py_ssize_t old_count = Py_REFCNT(obj);
+        for (int j = 0; j < 10000; j++) {
+            Py_DECREF(obj);
+        }
+        Py_ssize_t current_count = Py_REFCNT(obj);
+        if (old_count != current_count) {
+            PyErr_SetString(PyExc_RuntimeError, "Reference count should not be changed.");
+            return NULL;
+        }
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]:
```suggestion
        assert(old_count == current_count);
```

**Review by erlend-aasland — COMMENTED** [review]
I left some more comments on `test_immortal_bool`. The same comments applies to the rest of the test functions.

**Comment by corona10** [comment]
@erlend-aasland @JelleZijlstra @iritkatriel 
If there is no major objection, I prefer to merge this PR by next week.
I expect that if the more immortal type is added, more test cases will be added here.

File: Modules/_testcapi/immortal.c [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    Py_ssize_t n = Py_ARRAY_LENGTH(objects);
+    for (Py_ssize_t i = 0; i < n; i++) {
+        PyObject* obj = objects[i];
+        assert(_Py_IsImmortal(obj));
+        Py_ssize_t old_count = Py_REFCNT(obj);
+        for (int j = 0; j < 10000; j++) {
+            Py_DECREF(obj);
+        }
+        Py_ssize_t current_count = Py_REFCNT(obj);
+        assert(old_count == current_count);
```
Comment (erlend-aasland) [inline Modules/_testcapi/immortal.c]: You might want to separate this out as a helper function, since this part is repeated for all `test_*()` functions.

Comment (corona10) [reply Modules/_testcapi/immortal.c]: Done!

**Comment by corona10** [comment]
Merge this PR, if there needs some improvement, please let me know!

**Pull request merged**